test/core/end2end/inproc_callback_test.cc

```diff
@@ -0,0 +1,462 @@
+  void Run(bool ok) override {
+    gpr_log(GPR_DEBUG, "CQ shutdown notification invoked");
+    gpr_mu_lock(&mu_);
+    done_ = true;
+    gpr_cv_broadcast(&cv_);
+    gpr_mu_unlock(&mu_);
+  }
+  void Wait(gpr_timespec deadline) {
+    gpr_mu_lock(&mu_);
+    while (!done_ && !gpr_cv_wait(&cv_, &mu_, deadline)) {
+    }
+    gpr_mu_unlock(&mu_);
+  }
+
+ private:
+  bool done_;
+  gpr_mu mu_;
+  gpr_cv cv_;
+};
+
+ShutdownCallback* g_shutdown_callback;
+}  // namespace
+
+static gpr_mu tags_mu;
+static gpr_cv tags_cv;
+const size_t kAvailableTags = 4;
+bool tags[kAvailableTags];
+bool tags_valid[kAvailableTags];
+bool tags_expected[kAvailableTags];
+bool tags_needed[kAvailableTags];
+
+static void expect_tag(intptr_t tag, bool ok) {
+  size_t idx = static_cast<size_t>(tag);
+  GPR_ASSERT(idx < kAvailableTags);
+  tags_needed[idx] = true;
+  tags_expected[idx] = ok;
+}
+
+static void verify_tags(gpr_timespec deadline) {
+  bool done = false;
+
+  gpr_mu_lock(&tags_mu);
+  while (!done) {
+    done = gpr_time_cmp(gpr_now(GPR_CLOCK_MONOTONIC), deadline) > 0;
+    for (size_t i = 0; i < kAvailableTags; i++) {
+      if (tags_needed[i]) {
+        if (tags_valid[i]) {
+          gpr_log(GPR_DEBUG, "Verifying tag %d", static_cast<int>(i));
+          if (tags[i] != tags_expected[i]) {
+            gpr_log(GPR_ERROR, "Got wrong result (%d instead of %d) for tag %d",
+                    tags[i], tags_expected[i], static_cast<int>(i));
+          }
+          tags_valid[i] = false;
+          tags_needed[i] = false;
+        } else if (done) {
+          gpr_log(GPR_ERROR, "Didn't get tag %d", static_cast<int>(i));
+        }
+      }
+    }
+    bool empty = true;
+    for (size_t i = 0; i < kAvailableTags; i++) {
+      if (tags_needed[i]) {
+        empty = false;
+      }
+    }
+    done = done || empty;
+    if (done) {
+      for (size_t i = 0; i < kAvailableTags; i++) {
+        if (tags_valid[i]) {
+          gpr_log(GPR_ERROR, "Got unexpected tag %d and result %d",
+                  static_cast<int>(i), tags[i]);
+        }
+        tags_valid[i] = false;
+      }
+    } else {
+      gpr_cv_wait(&tags_cv, &tags_mu, deadline);
+    }
+  }
+  gpr_mu_unlock(&tags_mu);
+}
+
+static grpc_end2end_test_fixture inproc_create_fixture(
+    grpc_channel_args* client_args, grpc_channel_args* server_args) {
+  grpc_end2end_test_fixture f;
+  inproc_fixture_data* ffd = static_cast<inproc_fixture_data*>(
+      gpr_malloc(sizeof(inproc_fixture_data)));
+  memset(&f, 0, sizeof(f));
+
+  f.fixture_data = ffd;
+  g_shutdown_callback = grpc_core::New<ShutdownCallback>();
+  f.cq =
+      grpc_completion_queue_create_for_callback(g_shutdown_callback, nullptr);
+  f.shutdown_cq = grpc_completion_queue_create_for_pluck(nullptr);
+
+  return f;
+}
+
+void inproc_init_client(grpc_end2end_test_fixture* f,
+                        grpc_channel_args* client_args) {
+  f->client = grpc_inproc_channel_create(f->server, client_args, nullptr);
```

Comment:
Right now, the poller is based around a model of the application yielding a thread to it. And that's not consistent with the callback API since there is no application action in this case that indicates that the application is waiting for a result.